### ISSUE TITLE
New Parameter Config Format

### ISSUE THREAD

**Issue by jakmeier** [near/nearcore#6710]
The current parameter configuration file is a JSON dump of the `RuntimeConfig` struct in Rust. This has a couple of disadvantages that can be grouped in two categories.

1. Changes are cumbersome in many ways.
    - We have to copy-paste the full config file.
    - Git merges can slide in errors.
    - It is not easy to see the differences between versions.
    - Renaming fields requires all old files to be changed as well.
2. Enumerating parameters is impossible, which leads to messy ways of working around that.
    - Gas profiles use a separate `enum Cost` (defined in [core/primitives-core/src/profile.rs](https://github.com/near/nearcore/blob/09b28466103f9223513201616042d2f33af6d2e0/core/primitives-core/src/profile.rs#L176-L180)). Currently there is a one-to-many relationship between those costs and parameters. It appears to have grown organically. I think a one-to-one relationship would be cleaner. Which means, if we can enumerate parameters directly, the profile's `Cost` becomes obsolete.
    - The estimator currently has no programmatic link between parameters and estimations. Again, there is an `enum Cost` but this time defined in [runtime/runtime-params-estimator/src/cost.rs](https://github.com/near/nearcore/blob/09b28466103f9223513201616042d2f33af6d2e0/runtime/runtime-params-estimator/src/cost.rs#L9). This started out as a one-to-one relation to parameters but is now just a list of mostly independent estimations. Each listed estimation may be used to estiamte zero, one, or more parameters.

# Requirements on a new format
Introducing a new format for parameter configuration should solve both classes of problems listed above. The current suggestion is to do it with these two properties:
1. Config files can be differential.
2. There is an iterable and conclusive list of all parameters.

# Suggested solution
Based on [previous discussions](https://github.com/near/nearcore/pull/6657), I suggest we use a flat list of parameters defined in a suitable format (see options below) and store it as a list of key-value pairs when parsing raw files. From the list of KV pairs, we can manually construct a `RuntimeConfig`.

This allows a simple diffing strategy as follows. We have an initial config file and diffs-only config files for each new version. We compute the list of KV-pairs for a specific version by reading the initial file and merge it in-order with the diffs. Merging can be described as, existing keys are overwritten, new keys are added.

To solve the enumeration problem, keys would have to be from a defined list. This can be done by defining a C-like `enum Parameter`. The parser checks that keys are in that list. (But not all parameters must be set on every version.)

# New format spec
*work-in-progress, please add to the discussion as to help define the best option here*
## Option 1: Custom tailored simple key-value pairs
- One line per parameter, where each line contains a parameter and a value, separated by a column (`:`).
- Keys are alphanumeric strings that must exactly match a defined parameter (in `snake_case`).
- Values can be anything, the config parser will store them as strings. Only when using the value will it be parsed to an actual type, such as `Gas` or `Rational<u64>`.
- Comment lines are allowed if they start with a `#`.
- Empty lines and lines with only white-space characters are ignored by the parser.
- White-space around keys and values are trimmed by the parser.
```
# Example new format
some_key_without_quotes: 7
another_key: 12
values_can_also_be_strings: value_also_unquoted

# We can also define separate parsing strategies
some_gas_cost: 1_000_000_000
pessimistic_gas_increase_ratio: 103 / 100
```
Optionally, we can have a slightly different format for diffs that makes adding a parameter look different from updating one:
```
# Update shows both old and new value
wasm_regular_op_cost: 822756 -> 2207874
# Insertion only has a new value
wasm_read_cached_trie_node: 2280000000
```

**Pros**
- Flexible format for values
- Our parser is the living spec, thus we can change details and keep the format tightly coupled to our needs.

**Cons**
- Non standard format

## Option 2: Use standard format and keep hierarchie
Using something like TOML, or mabe just a simple reverse-object notation (`wasm_config.limit_config.some_max`) that keeps the current structure of `RuntimeConfig`

**Pros**
- Well-known format
- Well-defined format
- Can use serde for parsing

**Cons**
- Not flat enough: Enumeration of parameters and `RuntimeConfig` would have to be solved separately. (Solvable with some boiler-plate code.)
- Changing field names or layout of `RuntimeConfig` still requires changes in old config files.
- More work to implement custom value parsers

## Option 3: Use standard format but still flatten the list

We could use the same strategy for keys as in option 1 but store it in a standard format.

**Pros**
- All pros from option 2

**Con**
- Overkill: We literally only store keys and values without structure.
- More work to implement custom value parsers

**Comment by matklad** [comment]
@jakmeier Do you thing this is done aftre we've switched to flat config? Seems done enough to me :)

**Comment by jakmeier** [comment]
Yeah Option 1 has been implemented pretty much as described here.
